test/state/mpt.cpp

```diff
@@ -0,0 +1,282 @@
+
+namespace evmone::state
+{
+namespace
+{
+/// The collection of nibbles (4-bit values) representing a path in a MPT.
+struct Path
+{
+    size_t length;  // TODO: Can be converted to uint8_t.
+    uint8_t nibbles[64]{};
+
+    explicit Path(bytes_view k) noexcept : length(2 * k.size())
+    {
+        assert(length <= 64);
+        size_t i = 0;
+        for (const auto b : k)
+        {
+            // static_cast is only needed in GCC <= 8.
+            nibbles[i++] = static_cast<uint8_t>(b >> 4);
+            nibbles[i++] = static_cast<uint8_t>(b & 0x0f);
+        }
+    }
+
+    [[nodiscard]] Path tail(size_t index) const
+    {
+        assert(index <= length);
+        Path p{{}};
+        p.length = length - index;
+        std::memcpy(p.nibbles, &nibbles[index], p.length);
+        return p;
+    }
+
+    [[nodiscard]] Path head(size_t size) const
+    {
+        assert(size < length);
+        Path p{{}};
+        p.length = size;
+        std::memcpy(p.nibbles, nibbles, size);
+        return p;
+    }
+
+    [[nodiscard]] bytes encode(bool extended) const
+    {
+        bytes bs;
+        const auto is_even = length % 2 == 0;
+        if (is_even)
+            bs.push_back(0x00);
+        else
+            bs.push_back(0x10 | nibbles[0]);
+        for (size_t i = is_even ? 0 : 1; i < length; ++i)
+        {
+            const auto h = nibbles[i++];
+            const auto l = nibbles[i];
+            assert(h <= 0x0f);
+            assert(l <= 0x0f);
+            bs.push_back(uint8_t((h << 4) | l));
+        }
+        if (!extended)
+            bs[0] |= 0x20;
+        return bs;
+    }
+};
+}  // namespace
+
+/// The MPT Node.
+///
+/// The implementation is based on StackTrie from go-ethereum.
+class MPTNode
+{
+    enum class Kind : uint8_t
+    {
+        leaf,
+        ext,
+        branch
+    };
+
+    static constexpr size_t num_children = 16;
+
+    Kind m_kind = Kind::leaf;
+    Path m_path{{}};
+    bytes m_value;
+    std::unique_ptr<MPTNode> children[num_children];
+
+    MPTNode(Kind kind, const Path& path, bytes&& value = {}) noexcept
+      : m_kind{kind}, m_path{path}, m_value{std::move(value)}
+    {}
+
+    /// Named constructor for an extended node.
+    static MPTNode ext(const Path& k, std::unique_ptr<MPTNode> child) noexcept
+    {
+        MPTNode node{Kind::ext, k};
+        node.children[0] = std::move(child);
+        return node;
+    }
+
+    static size_t diff_index(const Path& p1, const Path& p2) noexcept
+    {
+        assert(p1.length <= p2.length);
+        return static_cast<size_t>(
+            std::mismatch(p1.nibbles, p1.nibbles + p1.length, p2.nibbles).first - p1.nibbles);
+    }
+
+public:
+    MPTNode() = default;
+
+    /// Named constructor for a leaf node.
+    static MPTNode leaf(const Path& k, bytes&& v) noexcept { return {Kind::leaf, k, std::move(v)}; }
+
+    void insert(const Path& k, bytes&& v);
+
+    [[nodiscard]] hash256 hash() const;
+};
+
+MPT::MPT() noexcept = default;
```

Comment:
Cosmetic: maybe move `MPT` method definitions after `class MPTNode` and all `MPTNode` method definitions.